File: mm_control/src/mm_control/control.py
"""Python base controller."""
import numpy as np
import rospy

from geometry_msgs.msg import Twist
from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint


class MMController(object):
    """Base controller class."""
    def __init__(self):
        self.q = np.zeros(9)
        self.dq = np.zeros(9)
        self.u = np.zeros(9)

        self.joint_state_rec = False

        self.joint_states_sub = rospy.Subscriber(
            "/mm_joint_states", JointState, self.joint_states_cb
        )

        self.ur10_joint_vel_pub = rospy.Publisher(
            "/ur_driver/joint_speed", JointTrajectory, queue_size=1
        )

        self.rb_joint_vel_pub = rospy.Publisher(
            "/ridgeback_velocity_controller/cmd_vel", Twist, queue_size=1
        )

    def publish_joint_speeds(self, now):
        # TODO need to use the base input mapping in the controller
        # c = np.cos(self.q[2])
        # s = np.sin(self.q[2])
        # R_wb = np.array([[c, -s], [s, c]])

        base_twist = Twist()
        base_twist.linear.x = self.u[0]
        base_twist.linear.y = self.u[1]
        base_twist.angular.z = self.u[2]

        arm_traj_point = JointTrajectoryPoint()
        arm_traj_point.velocities = list(self.u[3:])

        arm_traj = JointTrajectory()
        arm_traj.header.stamp = now
        arm_traj.points.append(arm_traj_point)

        self.ur10_joint_vel_pub.publish(arm_traj)
        self.rb_joint_vel_pub.publish(base_twist)

    def joint_states_cb(self, msg):
        self.q = np.array(msg.position)
        self.dq = np.array(msg.velocity)
        self.joint_state_rec = True

File: mm_control/src/control.cpp
#include "mm_control/control.h"

#include <ros/ros.h>
#include <Eigen/Eigen>

#include <trajectory_msgs/JointTrajectory.h>
#include <sensor_msgs/JointState.h>
#include <geometry_msgs/Twist.h>

#include <mm_kinematics/kinematics.h>

#include "mm_control/util/messages.h"


namespace mm {

MMController::MMController() {}
MMController::~MMController() {}


bool MMController::init(ros::NodeHandle& nh, const double hz) {
    this->hz = hz;
    dt = 1.0 / hz;

    did_become_unsafe = false;

    q = JointVector::Zero();
    dq = JointVector::Zero();
    u = JointVector::Zero();

    joint_state_rec = false;
    last_joint_state_time = ros::Time::now().toSec();

    mm_joint_states_sub = nh.subscribe("/mm_joint_states", 1,
            &MMController::mm_joint_states_cb, this);

    ur10_joint_vel_pub = nh.advertise<trajectory_msgs::JointTrajectory>(
            "/ur_driver/joint_speed", 1);
    rb_joint_vel_pub = nh.advertise<geometry_msgs::Twist>(
            "/ridgeback_velocity_controller/cmd_vel", 1);
}


void MMController::publish_joint_speeds(const ros::Time& now) {
    trajectory_msgs::JointTrajectory traj_arm_msg;
    geometry_msgs::Twist twist_base_msg;

    joint_speed_msgs(u, traj_arm_msg, twist_base_msg);

    traj_arm_msg.header.stamp = now;

    ur10_joint_vel_pub.publish(traj_arm_msg);
    rb_joint_vel_pub.publish(twist_base_msg);
}


void MMController::mm_joint_states_cb(const sensor_msgs::JointState& msg) {
    last_joint_state_time = msg.header.stamp.toSec();
    for (int i = 0; i < mm::NUM_JOINTS; ++i) {
        q(i) = msg.position[i];
        dq(i) = msg.velocity[i];
    }
    joint_state_rec = true;
}

} // namespace mm
